File: client/ios/DivKit/Actions/DivActionHandler.swift
import Foundation
import LayoutKit
import Serialization
import VGSL

public final class DivActionHandler {
  public typealias ShowTooltipAction = (TooltipInfo) -> Void
  public typealias TrackVisibility = (_ logId: String, _ cardId: DivCardID) -> Void

  typealias PerformTimerAction = (
    _ cardId: DivCardID,
    _ timerId: String,
    _ action: DivTimerAction
  ) -> Void

  typealias UpdateCardAction = (DivCardUpdateReason) -> Void

  private let urlHandler: DivUrlHandler
  private let trackVisibility: TrackVisibility
  private let trackDisappear: TrackVisibility
  private let variablesStorage: DivVariablesStorage
  private let functionsStorage: DivFunctionsStorage?
  private let persistentValuesStorage: DivPersistentValuesStorage
  private let blockStateStorage: DivBlockStateStorage
  private let updateCard: UpdateCardAction
  private let reporter: DivReporter
  private let idToPath: IdToPath
  private let flags: DivFlagsInfo

  private let animatorActionHandler: AnimatorActionHandler
  private let arrayActionsHandler = ArrayActionsHandler()
  private let clearFocusActionHandler = ClearFocusActionHandler()
  private let copyToClipboardActionHandler = CopyToClipboardActionHandler()
  private let dictSetValueActionHandler = DictSetValueActionHandler()
  private let downloadActionHandler: DownloadActionHandler
  private let focusElementActionHandler = FocusElementActionHandler()
  private let scrollActionHandler: ScrollActionHandler
  private let setStateActionHandler: SetStateActionHandler
  private let setStoredValueActionHandler: SetStoredValueActionHandler
  private let setVariableActionHandler = SetVariableActionHandler()
  private let submitActionHandler: SubmitActionHandler
  private let timerActionHandler: TimerActionHandler
  private let tooltipActionHandler: TooltipActionHandler
  private let videoActionHandler = VideoActionHandler()

  /// Do not create `DivActionHandler`. Use the instance from `DivKitComponents`.
  @_spi(Legacy)
  public convenience init(
    stateUpdater: DivStateUpdater = DefaultDivStateManagement(),
    blockStateStorage: DivBlockStateStorage = DivBlockStateStorage(),
    patchProvider: DivPatchProvider,
    submitter: DivSubmitter,
    variablesStorage: DivVariablesStorage = DivVariablesStorage(),
    functionsStorage: DivFunctionsStorage? = nil,
    updateCard: @escaping (DivCardUpdateReason) -> Void,
    showTooltip: ShowTooltipAction? = nil,
    tooltipActionPerformer: TooltipActionPerformer? = nil,
    trackVisibility: @escaping TrackVisibility = { _, _ in },
    trackDisappear: @escaping TrackVisibility = { _, _ in },
    urlHandler: DivUrlHandler,
    persistentValuesStorage: DivPersistentValuesStorage = DivPersistentValuesStorage(),
    reporter: DivReporter? = nil
  ) {
    self.init(
      stateUpdater: stateUpdater,
      blockStateStorage: blockStateStorage,
      patchProvider: patchProvider,
      submitter: submitter,
      variablesStorage: variablesStorage,
      functionsStorage: functionsStorage,
      updateCard: updateCard,
      showTooltip: showTooltip,
      tooltipActionPerformer: tooltipActionPerformer,
      trackVisibility: trackVisibility,
      trackDisappear: trackDisappear,
      performTimerAction: { _, _, _ in },
      urlHandler: urlHandler,
      persistentValuesStorage: persistentValuesStorage,
      reporter: reporter ?? DefaultDivReporter(),
      idToPath: IdToPath(),
      animatorController: DivAnimatorController(),
      flags: .default
    )
  }

  init(
    stateUpdater: DivStateUpdater,
    blockStateStorage: DivBlockStateStorage,
    patchProvider: DivPatchProvider,
    submitter: DivSubmitter,
    variablesStorage: DivVariablesStorage,
    functionsStorage: DivFunctionsStorage?,
    updateCard: @escaping UpdateCardAction,
    showTooltip: ShowTooltipAction?,
    tooltipActionPerformer: TooltipActionPerformer?,
    trackVisibility: @escaping TrackVisibility,
    trackDisappear: @escaping TrackVisibility,
    performTimerAction: @escaping PerformTimerAction,
    urlHandler: DivUrlHandler,
    persistentValuesStorage: DivPersistentValuesStorage,
    reporter: DivReporter,
    idToPath: IdToPath,
    animatorController: DivAnimatorController,
    flags: DivFlagsInfo
  ) {
    self.urlHandler = urlHandler
    self.trackVisibility = trackVisibility
    self.trackDisappear = trackDisappear
    self.variablesStorage = variablesStorage
    self.functionsStorage = functionsStorage
    self.persistentValuesStorage = persistentValuesStorage
    self.blockStateStorage = blockStateStorage
    self.updateCard = updateCard
    self.reporter = reporter
    self.idToPath = idToPath
    self.flags = flags

    animatorActionHandler = AnimatorActionHandler(animatorController: animatorController)
    downloadActionHandler = DownloadActionHandler(
      patchProvider: patchProvider,
      updateCard: updateCard
    )
    scrollActionHandler = ScrollActionHandler(
      blockStateStorage: blockStateStorage,
      updateCard: updateCard
    )
    setStateActionHandler = SetStateActionHandler(stateUpdater: stateUpdater)
    setStoredValueActionHandler = SetStoredValueActionHandler(
      persistentValuesStorage: persistentValuesStorage
    )
    submitActionHandler = SubmitActionHandler(submitter: submitter)
    timerActionHandler = TimerActionHandler(performer: performTimerAction)
    tooltipActionHandler = TooltipActionHandler(
      performer: tooltipActionPerformer,
      showTooltip: showTooltip
    )
  }

  public func handle(
    params: UserInterfaceAction.DivActionParams,
    sender: AnyObject?
  ) {
    let action: DivActionBase? = switch params.source {
    case .visibility:
      parseAction(type: DivVisibilityActionTemplate.self, json: params.action)
    case .disappear:
      parseAction(type: DivDisappearActionTemplate.self, json: params.action)
    default:
      parseAction(type: DivActionTemplate.self, json: params.action)
    }
    guard let action else {
      return
    }

    handle(
      action,
      path: params.path,
      source: params.source,
      localValues: params.localValues,
      sender: sender
    )
  }

  @_spi(Internal)
  public func handle(
    _ action: DivActionBase,
    path: UIElementPath,
    source: UserInterfaceAction.DivActionSource,
    localValues: [String: AnyHashable] = [:],
    sender: AnyObject?
  ) {
    let cardId = path.cardId
    let expressionResolver = ExpressionResolver(
      functionsProvider: FunctionsProvider(
        persistentValuesStorage: persistentValuesStorage
      ),
      customFunctionsStorageProvider: { [weak functionsStorage] in
        functionsStorage?.getStorage(path: path, contains: $0)
      },
      variableValueProvider: { [unowned variablesStorage] in
        if let value = localValues[$0] {
          return value
        }
        let variableName = DivVariableName(rawValue: $0)
        return variablesStorage.getVariableValue(path: path, name: variableName)
      },
      errorTracker: reporter.asExpressionErrorTracker(cardId: cardId)
    )

    guard action.resolveIsEnabled(expressionResolver) else {
      return
    }

    let path = if let scopeId = action.scopeId {
      idToPath[path.cardId.path + scopeId] ?? path
    } else {
      path
    }
    let info = action.resolveInfo(expressionResolver, path: path, source: source)
    let context = DivActionHandlingContext(
      info: info,
      expressionResolver: expressionResolver,
      variablesStorage: variablesStorage,
      blockStateStorage: blockStateStorage,
      actionHandler: self,
      updateCard: updateCard
    )

    switch action.typed {
    case let .divActionAnimatorStart(action):
      animatorActionHandler.handle(action, context: context)
    case let .divActionAnimatorStop(action):
      animatorActionHandler.handle(action, context: context)
    case let .divActionArrayInsertValue(action):
      arrayActionsHandler.handle(action, context: context)
    case let .divActionArrayRemoveValue(action):
      arrayActionsHandler.handle(action, context: context)
    case let .divActionArraySetValue(action):
      arrayActionsHandler.handle(action, context: context)
    case .divActionClearFocus:
      clearFocusActionHandler.handle(context: context)
    case let .divActionCopyToClipboard(action):
      copyToClipboardActionHandler.handle(action, context: context)
    case let .divActionDictSetValue(action):
      dictSetValueActionHandler.handle(action, context: context)
    case let .divActionDownload(action):
      downloadActionHandler.handle(action, context: context)
    case let .divActionFocusElement(action):
      focusElementActionHandler.handle(action, context: context)
    case let .divActionHideTooltip(action):
      tooltipActionHandler.handle(action, context: context)
    case let .divActionScrollBy(action):
      scrollActionHandler.handle(action, context: context)
    case let .divActionScrollTo(action):
      scrollActionHandler.handle(action, context: context)
    case let .divActionSetVariable(action):
      setVariableActionHandler.handle(action, context: context)
    case let .divActionSetState(action):
      setStateActionHandler.handle(action, context: context)
    case let .divActionSetStoredValue(action):
      setStoredValueActionHandler.handle(action, context: context)
    case let .divActionShowTooltip(action):
      tooltipActionHandler.handle(action, context: context)
    case let .divActionSubmit(action):
      submitActionHandler.handle(action, context: context)
    case let .divActionTimer(action):
      timerActionHandler.handle(action, context: context)
    case .divActionUpdateStructure(_):
      break
    case let .divActionVideo(action):
      videoActionHandler.handle(action, context: context)
    case .none:
      handleUrl(action, context: context, sender: sender)
    }

    reporter.reportAction(cardId: cardId, info: info)

    if source == .visibility {
      trackVisibility(info.logId, cardId)
    } else if source == .disappear {
      trackDisappear(info.logId, cardId)
    }
  }

  private func handleUrl(
    _ action: DivActionBase,
    context: DivActionHandlingContext,
    sender: AnyObject?
  ) {
    guard let url = context.info.url else {
      return
    }

    if let intent = DivActionIntent(url: url) {
      let cardId = context.cardId
      switch intent {
      case let .showTooltip(id, multiple):
        let tooltipInfo = TooltipInfo(id: id, showsOnStart: false, multiple: multiple)
        tooltipActionHandler.showTooltip(tooltipInfo)
      case let .hideTooltip(id):
        tooltipActionHandler.hideTooltip(id: id)
      case let .download(patchUrl):
        downloadActionHandler.handle(
          url: patchUrl,
          context: context,
          onSuccessActions: action.downloadCallbacks?.onSuccessActions,
          onFailActions: action.downloadCallbacks?.onFailActions
        )
      case let .setState(divStatePath, lifetime):
        setStateActionHandler.handle(
          divStatePath: divStatePath,
          lifetime: lifetime,
          context: context
        )
      case let .setVariable(name, value):
        context.variablesStorage.update(
          path: context.path,
          name: DivVariableName(rawValue: name),
          value: value
        )
      case let .setCurrentItem(id, index):
        scrollActionHandler.scrollToItem(cardId: cardId, id: id, index: index, animated: true)
      case let .setNextItem(id, step, overflow):
        scrollActionHandler.scrollToNextItem(
          cardId: cardId,
          id: id,
          step: step,
          overflow: overflow,
          animated: true
        )
      case let .setPreviousItem(id, step, overflow):
        scrollActionHandler.scrollToNextItem(
          cardId: cardId,
          id: id,
          step: -step,
          overflow: overflow,
          animated: true
        )
      case let .scroll(id, mode):
        switch mode {
        case .start:
          scrollActionHandler.scrollToStart(cardId: cardId, id: id, animated: true)
        case .end:
          scrollActionHandler.scrollToEnd(cardId: cardId, id: id, animated: true)
        case let .forward(offset, overflow):
          scrollActionHandler.scrollToOffset(
            cardId: cardId,
            id: id,
            offset: offset,
            isRelative: true,
            overflow: overflow,
            animated: true
          )
        case let .backward(offset, overflow):
          scrollActionHandler.scrollToOffset(
            cardId: cardId,
            id: id,
            offset: -offset,
            isRelative: true,
            overflow: overflow,
            animated: true
          )
        case let .position(position):
          scrollActionHandler.scrollToOffset(
            cardId: cardId,
            id: id,
            offset: position,
            animated: true
          )
        }
      case let .video(id: id, action: action):
        context.blockStateStorage.setState(
          id: id,
          cardId: cardId,
          state: VideoBlockViewState(state: action == .play ? .playing : .paused)
        )
        context.updateCard(.state(cardId))
      case let .timer(timerId, action):
        timerActionHandler.handle(cardId: cardId, timerId: timerId, action: action)
      case let .setStoredValue(storedValue):
        persistentValuesStorage.set(value: storedValue)
      }
      return
    }

    let info = context.info
    if !flags.useUrlHandlerForVisibilityActions,
       info.source == .visibility || info.source == .disappear {
      return
    }

    urlHandler.handle(url, info: info, sender: sender)
  }

  private func parseAction<T: TemplateValue>(
    type _: T.Type,
    json: JSONObject
  ) -> T.ResolvedValue? {
    try? DivTemplates.empty.parseValue(
      type: T.self,
      from: json.makeDictionary() ?? [:]
    ).unwrap()
  }
}

File: client/android/compose/src/main/kotlin/com/yandex/div/compose/actions/DivActionHandler.kt
package com.yandex.div.compose.actions

import android.net.Uri
import com.yandex.div.compose.DivReporter
import com.yandex.div.compose.dagger.DivContextScope
import com.yandex.div.internal.actions.DivUntypedAction
import com.yandex.div.internal.actions.isDivAction
import com.yandex.div.json.expressions.Expression
import com.yandex.div2.DivAction
import com.yandex.div2.DivActionAnimatorStart
import com.yandex.div2.DivActionAnimatorStopTemplate
import com.yandex.div2.DivActionClearFocus
import com.yandex.div2.DivActionCopyToClipboard
import com.yandex.div2.DivActionDownload
import com.yandex.div2.DivActionFocusElement
import com.yandex.div2.DivActionHideTooltip
import com.yandex.div2.DivActionScrollBy
import com.yandex.div2.DivActionScrollTo
import com.yandex.div2.DivActionSetCursorPosition
import com.yandex.div2.DivActionSetState
import com.yandex.div2.DivActionSetStoredValue
import com.yandex.div2.DivActionShowTooltip
import com.yandex.div2.DivActionSubmit
import com.yandex.div2.DivActionTimer
import com.yandex.div2.DivActionTyped
import com.yandex.div2.DivActionVideo
import com.yandex.div2.DivSightAction
import org.json.JSONObject
import javax.inject.Inject

@DivContextScope
internal class DivActionHandler @Inject constructor(
    private val externalActionHandler: DivExternalActionHandler,
    private val reporter: DivReporter,
    private val arrayActionsHandler: ArrayActionsHandler,
    private val dictSetValueActionHandler: DictSetValueActionHandler,
    private val setVariableActionHandler: SetVariableActionHandler,
    private val updateStructureActionHandler: UpdateStructureActionHandler
) {

    fun handle(
        context: DivActionHandlingContext,
        actions: List<DivAction>,
        source: DivActionSource
    ) {
        actions.forEach { handle(context = context, action = it, source = source) }
    }

    fun handle(context: DivActionHandlingContext, action: DivAction, source: DivActionSource) {
        handle(
            context = context,
            action = DivActionBase(
                id = action.logId,
                isEnabled = action.isEnabled,
                payload = action.payload,
                source = source,
                typed = action.typed,
                url = action.url
            )
        )
    }

    fun handle(context: DivActionHandlingContext, action: DivSightAction) {
        handle(
            context = context,
            action = DivActionBase(
                id = action.logId,
                isEnabled = action.isEnabled,
                payload = action.payload,
                source = DivActionSource.VISIBILITY,
                typed = action.typed,
                url = action.url
            )
        )
    }

    private fun handle(context: DivActionHandlingContext, action: DivActionBase) {
        val expressionResolver = context.expressionResolver
        if (!action.isEnabled.evaluate(expressionResolver)) {
            return
        }

        action.typed?.let {
            handle(context = context, action = it, baseAction = action)
            return
        }

        val url = action.url?.evaluate(expressionResolver) ?: return
        if (url.isDivAction) {
            DivUntypedAction.parse(url)?.let {
                handle(context = context, action = it)
            }
        } else {
            externalActionHandler.handle(
                context = context,
                action = DivActionData(
                    id = action.id.evaluate(expressionResolver),
                    payload = action.payload,
                    source = action.source,
                    url = action.url.evaluate(expressionResolver)
                )
            )
        }
    }

    private fun handle(
        context: DivActionHandlingContext,
        action: DivActionTyped,
        baseAction: DivActionBase
    ) {
        when (action) {
            is DivActionTyped.AnimatorStart -> notSupported(DivActionAnimatorStart.TYPE)
            is DivActionTyped.AnimatorStop -> notSupported(DivActionAnimatorStopTemplate.TYPE)
            is DivActionTyped.ArrayInsertValue ->
                arrayActionsHandler.handle(context, action.value)

            is DivActionTyped.ArrayRemoveValue ->
                arrayActionsHandler.handle(context, action.value)

            is DivActionTyped.ArraySetValue ->
                arrayActionsHandler.handle(context, action.value)

            is DivActionTyped.ClearFocus -> notSupported(DivActionClearFocus.TYPE)
            is DivActionTyped.CopyToClipboard -> notSupported(DivActionCopyToClipboard.TYPE)
            is DivActionTyped.Custom ->
                externalActionHandler.handleCustomAction(
                    context = context,
                    action = DivCustomActionData(
                        id = baseAction.id.evaluate(context.expressionResolver),
                        payload = baseAction.payload,
                        source = baseAction.source
                    )
                )

            is DivActionTyped.DictSetValue ->
                dictSetValueActionHandler.handle(context, action.value)

            is DivActionTyped.Download -> notSupported(DivActionDownload.TYPE)
            is DivActionTyped.FocusElement -> notSupported(DivActionFocusElement.TYPE)
            is DivActionTyped.HideTooltip -> notSupported(DivActionHideTooltip.TYPE)
            is DivActionTyped.ScrollBy -> notSupported(DivActionScrollBy.TYPE)
            is DivActionTyped.ScrollTo -> notSupported(DivActionScrollTo.TYPE)
            is DivActionTyped.SetCursorPosition -> notSupported(DivActionSetCursorPosition.TYPE)
            is DivActionTyped.SetState -> notSupported(DivActionSetState.TYPE)
            is DivActionTyped.SetStoredValue -> notSupported(DivActionSetStoredValue.TYPE)
            is DivActionTyped.SetVariable ->
                setVariableActionHandler.handle(context, action.value)

            is DivActionTyped.ShowTooltip -> notSupported(DivActionShowTooltip.TYPE)
            is DivActionTyped.Submit -> notSupported(DivActionSubmit.TYPE)
            is DivActionTyped.Timer -> notSupported(DivActionTimer.TYPE)
            is DivActionTyped.UpdateStructure ->
                updateStructureActionHandler.handle(context, action.value)

            is DivActionTyped.Video -> notSupported(DivActionVideo.TYPE)
        }
    }

    private fun handle(
        context: DivActionHandlingContext,
        action: DivUntypedAction
    ) {
        when (action) {
            is DivUntypedAction.HideTooltip -> notSupported("hide_tooltip")
            is DivUntypedAction.SetState -> notSupported("set_state")
            is DivUntypedAction.SetStoredValue -> notSupported("set_stored_value")
            is DivUntypedAction.SetVariable ->
                setVariableActionHandler.handle(context, action)

            is DivUntypedAction.ShowTooltip -> notSupported("show_tooltip")
            is DivUntypedAction.Timer -> notSupported("timer")
            is DivUntypedAction.Video -> notSupported("video")
        }
    }

    private fun notSupported(name: String) {
        reporter.reportError("Action not supported: $name")
    }
}

private class DivActionBase(
    val id: Expression<String>,
    val isEnabled: Expression<Boolean>,
    val payload: JSONObject?,
    val source: DivActionSource,
    val typed: DivActionTyped?,
    val url: Expression<Uri>?,
)
